File: web/src/services/group-service.ts
/**
 * Group Service
 * Handles group discovery and operations
 */

import type { Group, NearbyGroupsRequest, NearbyGroupsResponse } from '../domain/group';
import { calculateDistance } from '../domain/group';
import { get, post, put } from './api';
import { getDatabase } from './db';

/**
 * Discovers nearby groups
 * @param request - Nearby groups request with location and radius
 * @returns Promise resolving to nearby groups with distances
 */
export async function discoverNearbyGroups(
  request: NearbyGroupsRequest
): Promise<NearbyGroupsResponse> {
  try {
    const response = await get<Array<{ group: Group; distance: number; activity: number }>>(
      `/groups/nearby?latitude=${request.latitude}&longitude=${request.longitude}&radius=${request.radius}`
    );

    return {
      groups: response.map((item) => item.group),
      distances: response.map((item) => item.distance),
    };
  } catch {
    // Fallback to cached groups if API fails
    return getCachedNearbyGroups(request);
  }
}

/**
 * Gets cached nearby groups from RxDB
 * @param request - Nearby groups request
 * @returns Cached groups with calculated distances
 */
async function getCachedNearbyGroups(
  request: NearbyGroupsRequest
): Promise<NearbyGroupsResponse> {
  const db = await getDatabase();

  // Get all groups from cache
  const allGroups = await db.groups.find().exec();
  const groups = allGroups.map((doc) => doc.toJSON() as Group);

  // Calculate distances and filter by radius
  const results = groups
    .map((group) => {
      const distance = calculateDistance(
        request.latitude,
        request.longitude,
        group.latitude,
        group.longitude
      );
      return { group, distance };
    })
    .filter((result) => result.distance <= request.radius)
    .sort((a, b) => a.distance - b.distance);

  return {
    groups: results.map((r) => r.group),
    distances: results.map((r) => r.distance),
  };
}

/**
 * Creates a new group
 * @param group - Group creation data
 * @returns Created group
 */
export async function createGroup(group: {
  name: string;
  type: Group['type'];
  latitude: number;
  longitude: number;
  region_code?: string;
}): Promise<Group> {
  const response = await post<Group>('/groups', group);

  // Store in RxDB
  const db = await getDatabase();
  await db.groups.upsert(response);

  return response;
}

/**
 * Gets a group by ID
 * @param groupId - Group ID
 * @returns Group or null if not found
 */
export async function getGroup(groupId: string): Promise<Group | null> {
  try {
    // Use path parameter instead of query parameter: /groups/{id}
    const response = await get<Group>(`/groups/${groupId}`);

    // Store in RxDB
    const db = await getDatabase();
    await db.groups.upsert(response);

    return response;
  } catch {
    // Try to get from cache
    const db = await getDatabase();
    const cached = await db.groups.findOne(groupId).exec();
    return cached ? (cached.toJSON() as Group) : null;
  }
}

/**
 * Group suggestion response
 */
export interface GroupSuggestion {
  suggested_name: string;
  suggested_type: Group['type'];
}

/**
 * Gets group name and type suggestions based on location
 * @param latitude - Latitude
 * @param longitude - Longitude
 * @returns Promise resolving to suggestion
 */
export async function suggestGroupNameAndType(
  latitude: number,
  longitude: number
): Promise<GroupSuggestion> {
  try {
    const response = await get<GroupSuggestion>(
      `/groups/suggest?latitude=${latitude}&longitude=${longitude}`
    );
    return response;
  } catch {
    // Fallback to default suggestion
    return {
      suggested_name: '',
      suggested_type: 'village',
    };
  }
}

/**
 * Updates a group's name
 * @param groupId - Group ID
 * @param name - New name
 * @returns Updated group
 */
export async function updateGroupName(groupId: string, name: string): Promise<Group> {
  const response = await put<Group>(`/groups/${groupId}`, { name });

  // Update in RxDB
  const db = await getDatabase();
  await db.groups.upsert(response);

  return response;
}

File: api/internal/service/group_service.go
package service

import (
	"context"
	"fmt"
	"sort"

	"nearby-msg/api/internal/domain"
	"nearby-msg/api/internal/infrastructure/database"
	"nearby-msg/api/internal/utils"
)

// GroupService handles group business logic
type GroupService struct {
	repo *database.GroupRepository
}

// NewGroupService creates a new group service
func NewGroupService(repo *database.GroupRepository) *GroupService {
	return &GroupService{repo: repo}
}

// CreateGroupRequest represents a group creation request
type CreateGroupRequest struct {
	Name            string           `json:"name"`
	Type            domain.GroupType `json:"type"`
	Latitude        float64          `json:"latitude"`
	Longitude       float64          `json:"longitude"`
	RegionCode      *string          `json:"region_code,omitempty"`
	CreatorDeviceID string           `json:"creator_device_id"`
}

// CreateGroup creates a new group
func (s *GroupService) CreateGroup(ctx context.Context, req CreateGroupRequest) (*domain.Group, error) {
	// Check if device already created a group
	existingGroup, err := s.repo.GetByCreatorDeviceID(ctx, req.CreatorDeviceID)
	if err != nil {
		return nil, fmt.Errorf("failed to check existing group: %w", err)
	}
	if existingGroup != nil {
		return nil, fmt.Errorf("device has already created a group")
	}

	// Create group
	groupID, err := utils.GenerateID()
	if err != nil {
		return nil, fmt.Errorf("failed to generate group ID: %w", err)
	}

	group := &domain.Group{
		ID:              groupID,
		Name:            req.Name,
		Type:            req.Type,
		Latitude:        req.Latitude,
		Longitude:       req.Longitude,
		RegionCode:      req.RegionCode,
		CreatorDeviceID: req.CreatorDeviceID,
	}

	if err := group.Validate(); err != nil {
		return nil, err
	}

	if err := s.repo.Create(ctx, group); err != nil {
		return nil, fmt.Errorf("failed to create group: %w", err)
	}

	return group, nil
}

// ValidRadius represents valid radius values in meters
var ValidRadius = map[float64]bool{
	500:  true,
	1000: true,
	2000: true,
}

// ValidateRadius validates that radius is one of the allowed values
func ValidateRadius(radius float64) error {
	if !ValidRadius[radius] {
		return fmt.Errorf("invalid radius: must be 500, 1000, or 2000 meters")
	}
	return nil
}

// NearbyGroupsRequest represents a request to find nearby groups
type NearbyGroupsRequest struct {
	Latitude  float64 `json:"latitude"`
	Longitude float64 `json:"longitude"`
	Radius    float64 `json:"radius"` // in meters: 500, 1000, or 2000
}

// NearbyGroupResponse represents a nearby group with distance and activity
type NearbyGroupResponse struct {
	Group    *domain.Group `json:"group"`
	Distance float64       `json:"distance"` // in meters
	Activity int           `json:"activity"` // message count in last 24 hours (simplified)
}

// FindNearbyGroups finds groups within a radius and calculates activity
func (s *GroupService) FindNearbyGroups(ctx context.Context, req NearbyGroupsRequest) ([]NearbyGroupResponse, error) {
	// Validate radius
	if err := ValidateRadius(req.Radius); err != nil {
		return nil, err
	}

	// Find nearby groups
	results, err := s.repo.FindNearby(ctx, req.Latitude, req.Longitude, req.Radius)
	if err != nil {
		return nil, fmt.Errorf("failed to find nearby groups: %w", err)
	}

	// Sort by distance
	sort.Slice(results, func(i, j int) bool {
		return results[i].Distance < results[j].Distance
	})

	// Convert to response format with activity
	responses := make([]NearbyGroupResponse, len(results))
	for i, result := range results {
		// TODO: Calculate actual activity (message count in last 24h)
		// For now, return 0 as placeholder
		activity := 0
		responses[i] = NearbyGroupResponse{
			Group:    result.Group,
			Distance: result.Distance,
			Activity: activity,
		}
	}

	return responses, nil
}

// GetGroup retrieves a group by ID
func (s *GroupService) GetGroup(ctx context.Context, groupID string) (*domain.Group, error) {
	group, err := s.repo.GetByID(ctx, groupID)
	if err != nil {
		return nil, fmt.Errorf("failed to get group: %w", err)
	}

	return group, nil
}

// UpdateGroupRequest represents a group update request
type UpdateGroupRequest struct {
	Name string `json:"name"`
}

// UpdateGroup updates a group (only name for now)
func (s *GroupService) UpdateGroup(ctx context.Context, groupID string, deviceID string, req UpdateGroupRequest) (*domain.Group, error) {
	// Get the group first
	group, err := s.repo.GetByID(ctx, groupID)
	if err != nil {
		return nil, fmt.Errorf("failed to get group: %w", err)
	}

	// Check if the device is the creator
	if group.CreatorDeviceID != deviceID {
		return nil, fmt.Errorf("only the creator can update the group")
	}

	// Validate new name
	if len(req.Name) < 1 || len(req.Name) > 100 {
		return nil, domain.ErrInvalidGroupName
	}

	// Update the name
	if err := s.repo.UpdateName(ctx, groupID, req.Name); err != nil {
		return nil, fmt.Errorf("failed to update group: %w", err)
	}

	// Return updated group
	updatedGroup, err := s.repo.GetByID(ctx, groupID)
	if err != nil {
		return nil, fmt.Errorf("failed to get updated group: %w", err)
	}

	return updatedGroup, nil
}

// GroupSuggestionRequest represents a request for group name/type suggestions
type GroupSuggestionRequest struct {
	Latitude  float64 `json:"latitude"`
	Longitude float64 `json:"longitude"`
}

// GroupSuggestionResponse represents suggested group name and type
type GroupSuggestionResponse struct {
	SuggestedName string           `json:"suggested_name"`
	SuggestedType domain.GroupType `json:"suggested_type"`
}

// SuggestGroupNameAndType suggests a group name and type based on location
// This is a simplified implementation - in production, you might use reverse geocoding APIs
func (s *GroupService) SuggestGroupNameAndType(ctx context.Context, req GroupSuggestionRequest) (*GroupSuggestionResponse, error) {
	// Check for nearby groups to infer type
	nearbyReq := NearbyGroupsRequest{
		Latitude:  req.Latitude,
		Longitude: req.Longitude,
		Radius:    2000, // Check within 2km
	}

	nearbyGroups, err := s.FindNearbyGroups(ctx, nearbyReq)
	if err != nil {
		// If error, default to "other" type
		return &GroupSuggestionResponse{
			SuggestedName: "Community Group",
			SuggestedType: domain.GroupTypeOther,
		}, nil
	}

	// If there are nearby groups, suggest the most common type
	if len(nearbyGroups) > 0 {
		typeCount := make(map[domain.GroupType]int)
		for _, group := range nearbyGroups {
			typeCount[group.Group.Type]++
		}

		// Find most common type
		maxCount := 0
		var suggestedType domain.GroupType = domain.GroupTypeOther
		for groupType, count := range typeCount {
			if count > maxCount {
				maxCount = count
				suggestedType = groupType
			}
		}

		// Generate name based on type
		suggestedName := generateGroupName(suggestedType, req.Latitude, req.Longitude)

		return &GroupSuggestionResponse{
			SuggestedName: suggestedName,
			SuggestedType: suggestedType,
		}, nil
	}

	// No nearby groups - default suggestion
	// In production, use reverse geocoding to get actual location name
	return &GroupSuggestionResponse{
		SuggestedName: "Community Group",
		SuggestedType: domain.GroupTypeVillage,
	}, nil
}

// generateGroupName generates a group name based on type and coordinates
// This is a simplified implementation - in production, use reverse geocoding
func generateGroupName(groupType domain.GroupType, lat, lon float64) string {
	// Use coordinates to create a simple identifier
	// Format: "Type - {lat},{lon}" (rounded to 4 decimal places)
	latStr := fmt.Sprintf("%.4f", lat)
	lonStr := fmt.Sprintf("%.4f", lon)

	typeLabel := map[domain.GroupType]string{
		domain.GroupTypeVillage:          "Village",
		domain.GroupTypeHamlet:           "Hamlet",
		domain.GroupTypeResidentialGroup: "Residential Group",
		domain.GroupTypeStreetBlock:      "Street Block",
		domain.GroupTypeWard:             "Ward",
		domain.GroupTypeCommune:          "Commune",
		domain.GroupTypeApartment:        "Apartment",
		domain.GroupTypeResidentialArea:  "Residential Area",
		domain.GroupTypeOther:            "Community",
	}

	label := typeLabel[groupType]
	if label == "" {
		label = "Community"
	}

	return fmt.Sprintf("%s - %s,%s", label, latStr, lonStr)
}
